### ISSUE TITLE
Connection resets from sidecar when deleting a pod

### ISSUE THREAD

**Issue by pkossyfas** [istio/istio#40383]
### Bug Description

We are trying to achieve zero downtime deployment but we are getting some 502 http errors the moment a pod is being deleted.

We have the following setup:
- Public ALB on instance mode (we are facing the issue when using NLB as well)
- Cilium CNI (datapath: encapsulation/ipam: cluster pool/tunnel: vxlan)
- istio-proxy sidecar version 1.14.1
- externalTrafficPolicy: Cluster on the respective service

The backend application supports graceful shutdown and we have added a preStop hook both on the application and the istio sidecar but the issue is that whenever the pods gets into terminating phase - while the SIGTERM signal has not yet reached the containers due to the preStop hook - we are getting connection reset causing the ALB to return 502 http errors.

Note that we are not getting any error when removing the istio-proxy sidecar.

Here is an example flow:
- at 16:47:41 (UTC+3) we are deleting a pod (poltest-server-74ff5987c9-66fwp)
- The moment we are deleting the pod we are getting a lot of the following error from the load test tool we are using (fortio):
```
16:47:41 W http_client.go:838> [8] Non ok http code 502 (HTTP/1.1 502)
```
- istio-proxy receives the SIGTERM signal 4 seconds later (due to the preStop hook). Relevant logs (level: info, UTC time):
```
2022-08-10T13:47:45.235633Z	info	Status server has successfully terminated
2022-08-10T13:47:45.235635Z	info	Agent draining Proxy
2022-08-10T13:47:45.235730Z	error	accept tcp [::]:15020: use of closed network connection
2022-08-10T13:47:45.236622Z	info	Graceful termination period is 5s, starting...
2022-08-10T13:47:46.698820Z	warn	Envoy proxy is NOT ready: server is not live, current state is: DRAINING
2022-08-10T13:47:48.699140Z	warn	Envoy proxy is NOT ready: server is not live, current state is: DRAINING
2022-08-10T13:47:50.237764Z	info	Graceful termination period complete, terminating remaining proxies.
2022-08-10T13:47:50.237815Z	warn	Aborted all epochs
2022-08-10T13:47:50.237823Z	warn	Aborting epoch 0
2022-08-10T13:47:50.237934Z	info	Epoch 0 aborted normally
2022-08-10T13:47:50.238018Z	info	Agent has successfully terminated
```
- using tcpdump (capturing the traffic on the relevant instance nodeport 32511) and wireshark we are getting RST packet:
![Screenshot from 2022-08-10 16-52-59](https://user-images.githubusercontent.com/44210251/183918987-1642ee66-1604-425a-afa3-82fdbe695732.png)

### Version

```prose
$ istioctl version
client version: 1.14.1
control plane version: 1.14.1
data plane version: 1.14.1 (93 proxies)

$ kubectl version --short
Client Version: v1.21.0
Server Version: v1.21.7
```

**Comment by howardjohn** [comment]
This is ALB -> Pod (with sidecar) directly right? No ingressgateway?

Are you able to test going to pod directly (not via ALB) and/or removing the sidecar from the pod and see if it reproduces?

**Comment by pkossyfas** [comment]
Right, ALB --> Pod (with sidecar). Actually, it is alb target group --> instance nodeport (so there can be another extra hop from instance to instance).
We have tested also with NLB (no ingressgateway there as well) and we are getting those connection resets as well.

On the other hand, there are no errors when testing with ELB (ELB --> nginx proxy pod --> application pod) and when testing locally by port-forwarding the service (sidecar enabled in both cases).

**Comment by howardjohn** [comment]
port-forwarding to the service bypasses the sidecar, can you try hitting thecluster IP directly? That would rule out any AWS LB specific things

**Comment by pkossyfas** [comment]
**Test using cluster IP (no errors):**
Testing by hitting the cluster IP we receiving no errors. Results:
```
fortio load -c 8 -qps 300 -t 1m -timeout 3s http://100.69.186.90:8088/ping
```
```
Fortio 1.33.0 running at 300 queries per second, 8->8 procs, for 1m0s: http://100.69.186.90:8088/ping
15:13:07 I httprunner.go:93> Starting http test for http://100.69.186.90:8088/ping with 8 threads at 300.0 qps and parallel warmup
Starting at 300 qps with 8 thread(s) [gomax 8] for 1m0s : 2250 calls each (total 18000)
15:14:07 I periodic.go:723> T007 ended after 1m0.002872392s : 2250 calls. qps=37.498204840940005
15:14:07 I periodic.go:723> T006 ended after 1m0.002881974s : 2250 calls. qps=37.498198852764325
15:14:07 I periodic.go:723> T004 ended after 1m0.002949746s : 2250 calls. qps=37.498156499380976
15:14:07 I periodic.go:723> T002 ended after 1m0.002956631s : 2250 calls. qps=37.49815219667954
15:14:07 I periodic.go:723> T001 ended after 1m0.002969738s : 2250 calls. qps=37.49814400561362
15:14:07 I periodic.go:723> T005 ended after 1m0.003091697s : 2250 calls. qps=37.49806778893852
15:14:07 I periodic.go:723> T003 ended after 1m0.003148185s : 2250 calls. qps=37.49803248760988
15:14:07 I periodic.go:723> T000 ended after 1m0.003167124s : 2250 calls. qps=37.498020651980674
Ended after 1m0.003192912s : 18000 calls. qps=299.98
Sleep times : count 17992 avg 0.024014967 +/- 0.0007295 min 0.00570817 max 0.025396916 sum 432.077287
Aggregated Function Time : count 18000 avg 0.0020312857 +/- 0.0006166 min 0.00111114 max 0.020600165 sum 36.5631434
# range, mid point, percentile, count
>= 0.00111114 <= 0.002 , 0.00155557 , 54.11, 9739
> 0.002 <= 0.003 , 0.0025 , 98.80, 8045
> 0.003 <= 0.004 , 0.0035 , 99.40, 108
> 0.004 <= 0.005 , 0.0045 , 99.52, 22
> 0.005 <= 0.006 , 0.0055 , 99.61, 16
> 0.006 <= 0.007 , 0.0065 , 99.73, 21
> 0.007 <= 0.008 , 0.0075 , 99.79, 12
> 0.008 <= 0.009 , 0.0085 , 99.89, 18
> 0.009 <= 0.01 , 0.0095 , 99.91, 3
> 0.01 <= 0.011 , 0.0105 , 99.92, 1
> 0.011 <= 0.012 , 0.0115 , 99.94, 4
> 0.012 <= 0.014 , 0.013 , 99.98, 7
> 0.016 <= 0.018 , 0.017 , 99.98, 1
> 0.02 <= 0.0206002 , 0.0203001 , 100.00, 3
# target 50% 0.00193255
# target 75% 0.0024675
# target 90% 0.00280311
# target 99% 0.00333333
# target 99.9% 0.00933333
Error cases : no data
15:14:07 I httprunner.go:190> [0] 100.69.186.90:8088 resolved to 100.69.186.90:8088
15:14:07 I httprunner.go:190> [1] 100.69.186.90:8088 resolved to 100.69.186.90:8088
15:14:07 I httprunner.go:190> [2] 100.69.186.90:8088 resolved to 100.69.186.90:8088
15:14:07 I httprunner.go:190> [3] 100.69.186.90:8088 resolved to 100.69.186.90:8088
15:14:07 I httprunner.go:190> [4] 100.69.186.90:8088 resolved to 100.69.186.90:8088
15:14:07 I httprunner.go:190> [5] 100.69.186.90:8088 resolved to 100.69.186.90:8088
15:14:07 I httprunner.go:190> [6] 100.69.186.90:8088 resolved to 100.69.186.90:8088
15:14:07 I httprunner.go:190> [7] 100.69.186.90:8088 resolved to 100.69.186.90:8088
Sockets used: 8 (for perfect keepalive, would be 8)
Uniform: false, Jitter: false
IP addresses distribution:
100.69.186.90:8088: 8
Code 200 : 18000 (100.0 %)
Response Header Sizes : count 18000 avg 172.00078 +/- 0.02788 min 172 max 173 sum 3096014
Response Body/Total Sizes : count 18000 avg 190.00078 +/- 0.02788 min 190 max 191 sum 3420014
All done 18000 calls (plus 8 warmup) 2.031 ms avg, 300.0 qps
```

**Test using Nodeport(we get connection resets):**
Testing using a random worker IP and the respective service nodeport we have the following results:
```
fortio load -c 8 -qps 300 -t 1m -timeout 3s http://10.1.86.24:32511/ping
```
```
Fortio 1.33.0 running at 300 queries per second, 8->8 procs, for 1m0s: http://10.1.86.24:32511/ping
15:18:14 I httprunner.go:93> Starting http test for http://10.1.86.24:32511/ping with 8 threads at 300.0 qps and parallel warmup
Starting at 300 qps with 8 thread(s) [gomax 8] for 1m0s : 2250 calls each (total 18000)
15:18:27 I http_client.go:806> [6] Closing dead socket 10.1.86.24:32511 (err read tcp 100.126.122.57:47814->10.1.86.24:32511: read: connection reset by peer at first read)
15:18:27 I http_client.go:806> [1] Closing dead socket 10.1.86.24:32511 (err read tcp 100.126.122.57:47824->10.1.86.24:32511: read: connection reset by peer at first read)
15:19:14 I periodic.go:723> T005 ended after 1m0.001684947s : 2250 calls. qps=37.49894693769757
15:19:14 I periodic.go:723> T000 ended after 1m0.001825608s : 2250 calls. qps=37.49885902971608
15:19:14 I periodic.go:723> T007 ended after 1m0.001830348s : 2250 calls. qps=37.49885606739658
15:19:14 I periodic.go:723> T004 ended after 1m0.001858883s : 2250 calls. qps=37.49883823411812
15:19:14 I periodic.go:723> T006 ended after 1m0.001844539s : 2250 calls. qps=37.49884719856479
15:19:14 I periodic.go:723> T003 ended after 1m0.001895781s : 2250 calls. qps=37.49881517431117
15:19:14 I periodic.go:723> T002 ended after 1m0.001907321s : 2250 calls. qps=37.49880796226831
15:19:14 I periodic.go:723> T001 ended after 1m0.001930662s : 2250 calls. qps=37.49879337507642
Ended after 1m0.001948527s : 18000 calls. qps=299.99
Sleep times : count 17992 avg 0.024508046 +/- 0.0006779 min 0.012145583 max 0.025605314 sum 440.948767
Aggregated Function Time : count 18000 avg 0.0015282839 +/- 0.0005055 min 0.000883541 max 0.014081757 sum 27.5091095
# range, mid point, percentile, count
>= 0.000883541 <= 0.001 , 0.00094177 , 0.02, 4
> 0.001 <= 0.002 , 0.0015 , 94.96, 17088
> 0.002 <= 0.003 , 0.0025 , 98.97, 722
> 0.003 <= 0.004 , 0.0035 , 99.43, 84
> 0.004 <= 0.005 , 0.0045 , 99.56, 23
> 0.005 <= 0.006 , 0.0055 , 99.65, 16
> 0.006 <= 0.007 , 0.0065 , 99.78, 24
> 0.007 <= 0.008 , 0.0075 , 99.86, 14
> 0.008 <= 0.009 , 0.0085 , 99.88, 3
> 0.009 <= 0.01 , 0.0095 , 99.96, 14
> 0.01 <= 0.011 , 0.0105 , 99.99, 6
> 0.012 <= 0.014 , 0.013 , 99.99, 1
> 0.014 <= 0.0140818 , 0.0140409 , 100.00, 1
# target 50% 0.00152645
# target 75% 0.00178979
# target 90% 0.0019478
# target 99% 0.00307143
# target 99.9% 0.00928571
Error cases : no data
15:19:14 I httprunner.go:190> [0] 10.1.86.24:32511 resolved to 10.1.86.24:32511
15:19:14 I httprunner.go:190> [1] 10.1.86.24:32511 resolved to 10.1.86.24:32511
15:19:14 I httprunner.go:190> [2] 10.1.86.24:32511 resolved to 10.1.86.24:32511
15:19:14 I httprunner.go:190> [3] 10.1.86.24:32511 resolved to 10.1.86.24:32511
15:19:14 I httprunner.go:190> [4] 10.1.86.24:32511 resolved to 10.1.86.24:32511
15:19:14 I httprunner.go:190> [5] 10.1.86.24:32511 resolved to 10.1.86.24:32511
15:19:14 I httprunner.go:190> [6] 10.1.86.24:32511 resolved to 10.1.86.24:32511
15:19:14 I httprunner.go:190> [7] 10.1.86.24:32511 resolved to 10.1.86.24:32511
Sockets used: 10 (for perfect keepalive, would be 8)
Uniform: false, Jitter: false
IP addresses distribution:
10.1.86.24:32511: 8
Code 200 : 18000 (100.0 %)
Response Header Sizes : count 18000 avg 259 +/- 0 min 259 max 259 sum 4662000
Response Body/Total Sizes : count 18000 avg 277 +/- 0 min 277 max 277 sum 4986000
All done 18000 calls (plus 8 warmup) 1.528 ms avg, 300.0 qps
```
Checking the results fortio reports 100.0% success code (http 200), so it seems that in this case the connection resets do not drop any inflight request. (in the case of ALB those connection resets are causing a lot of 502 http errors) 

15:18:27 is the time we deleted a pod.

For the record, here is the service details:
```
$ kubectl describe svc poltest-server
Name:                     poltest-server
Namespace:                polykarpos
Labels:                   app=poltest-server
                          app.kubernetes.io/instance=poltest-server
                          app.kubernetes.io/managed-by=Helm
                          app.kubernetes.io/name=poltest-server
                          domain=platform
                          helm.sh/chart=poltest-server
                          team=exampleTeam
                          version=1.0
Annotations:              meta.helm.sh/release-name: poltest-server
                          meta.helm.sh/release-namespace: polykarpos
Selector:                 app.kubernetes.io/instance=poltest-server,app.kubernetes.io/name=poltest-server
Type:                     NodePort
IP Family Policy:         SingleStack
IP Families:              IPv4
IP:                       100.69.186.90
IPs:                      100.69.186.90
Port:                     http  8088/TCP
TargetPort:               8088/TCP
NodePort:                 http  32511/TCP
Endpoints:                100.125.85.46:8088,100.126.79.243:8088,100.127.94.37:8088
Session Affinity:         None
External Traffic Policy:  Cluster
Events:                   <none>
```

To my understanding, the moment a pod goes into terminating state, k8s endpoint controller removes the respective pod ip from the service's endpoints without dropping any connection, since both the proxy sidecar and the application are still alive due to the preStop hook. And in the above tests fortio is using keepalives so we should not observe the above connection resets.

I do not know the istio internals, so I was wondering if istiod monitors the service's endpoints list and performs any action following a change on those endpoints.
